### DISCUSSION TITLE
"null" car_id for multiple graphs

### DISCUSSION THREAD

**Discussion by arcastro** [teslamate-org/teslamate#4180] [Ideas]
Hello,

Many of my graphs on grafana show a "null" car_id.

Is this expected?

For example, here is my "mileage" graph

![image](https://github.com/user-attachments/assets/e24e8730-84b1-46c5-bbd7-dbfbead2389a)

What is the car_id expected to be? VIN? custom name?

I appreciate the help

**Comment by cwanja** [comment]
`car_id` should be a unique value assigned to a car record in the database. It's odd you have no ID. 

From the Explore part of Grafana, can you run a `select * from cars;` and share? Mask any data you'd like.

**Reply by arcastro** [reply]
I do see one row returned

![image](https://github.com/user-attachments/assets/89b9e1e8-db24-4a35-98f8-ab018cd839e6)

Does this mean the $car_id should be simply `1`?

**Comment by brianmay** [comment]
I have seen this happen when I don't have a name set for the car.

**Reply by arcastro** [reply]
Thank you @brianmay , where would I set the car name to try it?

**Reply by brianmay** [reply]
It is a config option in the car.

**Reply by cwanja** [reply]
Sounds like a TeslaMate bug we need to investigate @brianmay.

**Reply by brianmay** [reply]
My guess is that grafana assumes there is a car name set, and defaults to null if there isn't. But I haven't really looked.

**Reply by cwanja** [reply]
Grafana is just reading the database. If the record is getting crated with a null value (how, figured ID is a required field) at entry due to no set name, that should be fixed in TeslaMate.

**Reply by brianmay** [reply]
For me it displays the car *name* not the car id.

**Reply by cwanja** [reply]
> For me it displays the car *name* not the car id.

Right. But before any of that, the record has to get entered into the database. If the name is missing (or not set) from the Tesla API, curious if it's placing a null value for `car_id`. Which also seems odd, because I imagine an ID should be an automatic value and required field. Does that make sense?

**Reply by DrMichael** [reply]
I'd say, two possibilities: either set the name in teslamate to something like "unknown" or handle the null name in all places in Grafana...

Btw, what happens, if one would set or change the name?

**Reply by cwanja** [reply]
> Btw, what happens, if one would set or change the name?

It changes in the database on the next call to the API. But it seems, and we would have to test. If you have not yet 'named' your car, when you connect TeslaMate it stores a null for car_id. Wonder if we just say if null, set 'Tesla' as the name.

**Reply by DrMichael** [reply]
You mean by teslamate in the database? Or "no name"? Sounds good to me.
I would not address that in Grafana, that would be in probably almost every dashboard.

**Reply by arcastro** [reply]
The existing grafana dashboard variable is populated from this db query:

```
SELECT name AS __text, id AS __value FROM cars ORDER BY display_priority ASC, name ASC;
```

![image](https://github.com/user-attachments/assets/f996e4bb-f3c4-46db-8e05-cab4b22402c0)


What if we change it to something like this?

```
SELECT COALESCE(name, CONCAT('VIN ', vin)) AS __text, id AS __value FROM cars ORDER BY display_priority ASC, name ASC;
```

When the name is empty, it would instead show "VIN 123456789ABC", I think this is better vehicle identification than displaying "Tesla" or "No Name". It also prevents us from forcibly storing an arbitrary name to the database. I think the database should represent the truth, and the truth is there is no name.

**Reply by arcastro** [reply]
Quick side note, 

I added a name to my vehicle in the console settings , and it looks like the new name hasn't been updated in the teslamate database so my graphs still show up as null name. But that was only 30 minutes ago, or so. Do I need to wait longer for teslamate to pick up the vehicle rename, or is the `cars` database only populated once and never updated? Thank you everyone who has chimed in :)

EDIT: looks like I just needed to wait. It shows up now.

**Reply by DrMichael** [reply]
> When the name is empty, it would instead show "VIN 123456789ABC", I think this is better vehicle identification than displaying "Tesla" or "No Name". It also prevents us from forcibly storing an arbitrary name to the database. I think the database should represent the truth, and the truth is there is no name.

As I said, that would have to be changed in each dashboard. I would prefer to ix that in the elixir code of Teslamate.
I like the idea with the VIN though...

**Reply by cwanja** [reply]
Agreed with @DrMichael - the better fix is at runtime in the application / written to the database. Dashboards should just read from the source of truth, the database.

**Reply by cwanja** [reply]
I did misspeak earlier - there is no "`car_id`", that is just the variable as @arcastro pointed out. It is just reading the `name` field from the `cars` entity. Just odd that Tesla does not default to "Tesla" in the API when the car has not been named.

**Reply by arcastro** [reply]
Thank you again @DrMichael and @cwanja , 
I'll defer to your decision to default the name directly in the database. Sounds good to me 👍

Would still like to propose defaulting to VIN, as opposed to just "Tesla" if you both agree - I'm assuming Teslamate supports tracking multiple vehicles and it would help distinguish between them if they are both un-named.

**Reply by arcastro** [reply]
What's the next step here? Should we close the discussion and move this to an issue, so it doesn't get lost?

**Reply by cwanja** [reply]
It is not really an "issue", everything is working as intended. This would be an enhancement which would have to be picked up by anyone who could develop the capability. I have gone ahead and changed it to an Idea.

**Reply by arcastro** [reply]
> It is not really an "issue", everything is working as intended.

I think displaying `null` for unnamed vehicles would be viewed as a bug by most people 😉, although I agree it's low-priority since it's only cosmetic and there is an easy mitigation (naming your vehicle).

I'd be happy to submit a pull request updating [the locations here](https://github.com/search?q=repo%3Ateslamate-org%2Fteslamate+%22__value+FROM+cars%22&type=code). I still think this is a presentation issue that should be handled at the presentation layer, rather than an issue with the data in the database (since the data in the database is actually correct). 

Otherwise, I'm okay to let this sit for now. Ty for updating the discussion category :)

**Reply by cwanja** [reply]
> I think displaying null for unnamed vehicles would be viewed as a bug by most people 😉

But an unnamed vehicle _is_ `null` right? At this point, we would be arguing over semantics 😁

> I'd be happy to submit a pull request updating [the locations here](https://github.com/search?q=repo%3Ateslamate-org%2Fteslamate+%22__value+FROM+cars%22&type=code).

By all means, go for it! We love contributors. The lovely team here will review and give feedback and the discussion can move from here to that PR. And if the time comes where a revision makes its way into the code that writes to the database, your changes could be undone.

**Reply by arcastro** [reply]
> But an unnamed vehicle _is_ `null` right?

😄 you got me there!
And ty again I'll follow up soon

**Reply by arcastro** [reply]
opened here: https://github.com/teslamate-org/teslamate/pull/4198

**Comment by arcastro** [comment]
Resolved in https://github.com/teslamate-org/teslamate/pull/4198
And released in https://github.com/teslamate-org/teslamate/releases/tag/v1.31.0
